src/vcf/assembly_checker.cpp

```diff
@@ -0,0 +1,101 @@
+/**
+ * Copyright 2014-2018 EMBL - European Bioinformatics Institute
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include <set>
+#include <vector>
+
+#include <boost/algorithm/string/predicate.hpp>
+
+#include "bioio/bioio.hpp"
+#include "util/logger.hpp"
+#include "util/stream_utils.hpp"
+#include "vcf/assembly_checker.hpp"
+#include "vcf/vcf_fasta_variant.hpp"
+
+namespace ebi
+{
+  namespace vcf
+  {
+    namespace assembly_checker
+    {
+
+      bool check_vcf_ref(std::istream &vcf_input, std::istream &fasta_input, std::istream &fasta_index_input,
+                         std::ostream &problem_lines_output)
+      {
+          std::vector<char> vector_line;
+          vector_line.reserve(default_line_buffer_size);
+
+          std::set<std::string> absent_chromosomes;
+          size_t num_matches = 0;
+          size_t num_variants = 0;
+
+          // Reading FASTA index, and querying FASTA file
+          auto index = bioio::read_fasta_index(fasta_index_input);
+
+          // Reading VCF file
+          while (util::readline(vcf_input, vector_line).size() != 0) {
```

Comment:
try to think and extract functions from inside the loop. experiment with ideas like:
- function for checks on preconditions
- function for checking a VcfVariant
- function for keeping track of counts

try a structure where each function can be described (and named) as doing just one thing